pxteditor/code-validation/runValidatorPlanAsync.ts

```diff
@@ -5,33 +5,48 @@ import { validateBlocksInSetExist } from "./validateBlocksInSetExist";
-    // TBD if it's faster to run in parallel without short-circuiting once the threshold is reached, or if it's faster to run sequentially and short-circuit.
+export function runValidatorPlanAsync(usedBlocks: Blockly.Block[], plan: pxt.blocks.ValidatorPlan, planBank: pxt.blocks.ValidatorPlan[]): boolean {
     const startTime = Date.now();
+    let checksSucceeded = 0;
+    let successfulBlocks: Blockly.Block[] = [];
 
-    const checkRuns = pxt.Util.promisePoolAsync(maxConcurrentChecks, plan.checks, async (check: pxt.blocks.ValidatorCheckBase): Promise<boolean> => {
+    for (const check of plan.checks) {
+        let checkPassed = false;
         switch (check.validator) {
             case "blocksExist":
-                return runBlocksExistValidation(usedBlocks, check as pxt.blocks.BlocksExistValidatorCheck);
+                [successfulBlocks, checkPassed] = [...runBlocksExistValidation(usedBlocks, check as pxt.blocks.BlocksExistValidatorCheck)];
+                break;
             case "blockCommentsExist":
-                return runValidateBlockCommentsExist(usedBlocks, check as pxt.blocks.BlockCommentsExistValidatorCheck);
+                checkPassed = runValidateBlockCommentsExist(usedBlocks, check as pxt.blocks.BlockCommentsExistValidatorCheck);
+                break;
             case "specificBlockCommentsExist":
-                return runValidateSpecificBlockCommentsExist(usedBlocks, check as pxt.blocks.SpecificBlockCommentsExistValidatorCheck);
+                checkPassed = runValidateSpecificBlockCommentsExist(usedBlocks, check as pxt.blocks.SpecificBlockCommentsExistValidatorCheck);
+                break;
             case "blocksInSetExist":
-                return runBlocksInSetExistValidation(usedBlocks, check as pxt.blocks.BlocksInSetExistValidatorCheck);
+                [successfulBlocks, checkPassed] = [...runBlocksInSetExistValidation(usedBlocks, check as pxt.blocks.BlocksInSetExistValidatorCheck)];
+                break;
             default:
                 pxt.debug(`Unrecognized validator: ${check.validator}`);
-                return false;
+                checkPassed = false;
+                break;
         }
-    });
 
-    const results = await checkRuns;
-    const successCount = results.filter((r) => r).length;
-    const passed = successCount >= plan.threshold;
+        if (checkPassed && check.childValidatorPlans) {
+            for (const planName of check.childValidatorPlans) {
+                let timesPassed = 0;
+                for (const parentBlock of successfulBlocks) {
+                    const blocksToUse = parentBlock.getChildren(true);
+                    const childPlan = planBank.find((plan) => plan.name === planName);
+                    const childPassed = runValidatorPlanAsync(blocksToUse, childPlan, planBank);
+                    timesPassed += childPassed ? 1 : 0;
+                }
+                checkPassed = timesPassed > 0;
```

Comment:
Potential bug here. In the course of iterating over child validator plans, `checkPassed` can become false then true again with no consequences. Only the last-assigned value of `checkPassed` is used when computing `checksSucceeded` below.

Reply:
Yeah, good catch. I think "anding" this will do the trick.. 

If we have any of the children validators fail, then the whole validator plan should fail, so, if at any moment, `checkPassed` is false, then that false value should be maintained.